lib/omni-sharp-server/omni-sharp-server.ts

```diff
@@ -60,54 +61,138 @@ atom.emitter.on("omni-sharp-server:state-change", (state) => {
 class OmniSharpServerInstance {
     public packageDir = atom.packages.getPackageDirPaths()[0];
     public location = OmnisharpLocation;
-    public child;
+    public child: child_process.ChildProcess;
     public port;
+    private _seq = 0;
 
     public start() {
         var executablePath, useMono;
         useMono = process.platform !== "win32";
         executablePath = this.location;
-        return findFreePort((err, port) => {
+        findFreePort((err, port) => {
             var ref;
             if (err) {
                 return console.error("error finding freeport: ", err);
             }
             this.port = port;
-            var serverArguments = ["-s", typeof atom !== "undefined" && atom !== null ? (ref = atom.project) != null ? ref.getPaths()[0] : void 0 : void 0, "-p", port];
+
+
+            var path = atom && atom.project && atom.project.getPaths()[0] || void 0;
+            var serverArguments = ["--stdio", "-s", path, "-p", port, "--hostPID", process.pid];
             this.child = spawn(executablePath, serverArguments);
             atom.emitter.emit("omni-sharp-server:start", this.child.pid, port);
             atom.emitter.emit("omni-sharp-server:state-change", "loading");
-            this.child.stdout.on('data', this.out);
-            this.child.stderr.on('data', this.err);
+            this.child.stdout.on('data', this.serverStart);
+            this.child.stderr.on('data', this.serverErr);
+            this.child.stdout.on('data', this.serverResponse);
             this.child.on('close', this.close);
-            return this.child.on('error', this.err);
+            this.child.on('error', this.serverErr);
         })
     }
 
-    public out = (data) => {
+    private serverStart = (data) => {
         var s = data.toString();
 
-        var ref = s.match(/Solution has finished loading/);
-        if ((ref != null ? ref.length : 0) > 0) {
-            atom.emitter.emit("omni-sharp-server:ready", this.child.pid);
-            atom.emitter.emit("omni-sharp-server:state-change", "ready");
-        }
-
         var ref = s.match(/Detected an OmniSharp instance already running on port/);
         if ((ref != null ? ref.length : 0) > 0) {
             atom.emitter.emit("omni-sharp-server:error");
             atom.emitter.emit("omni-sharp-server:state-change", "error");
             this.stop();
         }
+    }
+
+    private _outstandingRequests: { [seq: number]: { resolve: (thenable: any | Promise.Thenable<any>) => void; reject: (error: any) => void; } } = {}
+    private _partialResult: string;
+
+    // TODO: Review if this is the best way...
+    private serverResponse = (data: any) => {
+        var result :string = data.toString();
+        if (this._partialResult) {
+            result = this._partialResult + result;
+        }
+
+        var results = result.split('\n');
```

Comment:
readline https://nodejs.org/api/readline.html would probably simplify this code.